### PR TITLE
feature(api) add API blueprint workflow

### PR DESCRIPTION
**Pull request by jmaupetit**
- [x] Add POC API blueprint specification with a single `/status/version` endpoint
- [x] Implement this API view with DRF
- [x] Add & configure dredd
- [x] Integrate dredd to the CI workflow
- [x] Check OpenAPI schema
- [x] Add docs
- [x] Drop Django 1.8 support

### CODE REVIEW

File: fonzie/urls/schema.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+from django.conf.urls import url
+
+from ..apps import FonzieConfig
+from ..views.schema import FonzieSchemaView
+
+app_name = FonzieConfig.name
+urlpatterns = [
+    url(
+        r'^$',
+        FonzieSchemaView.as_view(title="Fonzie, an Open edX CRUD API"),
```
Comment (sampaccoud) [inline fonzie/urls/schema.py]: no more CRUD ?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "fonzie",
+  "version": "0.1.0",
+  "description": "Open edX CRUD API",
```
Comment (sampaccoud) [inline package.json]: no more CRUD?

File: README.rst [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
+in touch if you have any question related to our implementation or design
+decisions.
 
-edX CRUD API
+Overview
+--------
 
-Overview (please modify)
-------------------------
+Fonzie is an HTTP API over the `Open edX <https://open.edx.org/>`_ project. It
```
Comment (sampaccoud) [inline README.rst:15]: Rest ?

File: README.rst [hunk 4]
```diff
@@ -36,21 +41,25 @@ How To Contribute
 
 Contributions are very welcome.
 
-Please read `How To Contribute <https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst>`_ for details.
+Please read `How To Contribute
+<https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst>`_ for
+details.
```
Comment (sampaccoud) [inline README.rst]: I would say we remove this until we add guide lines on how to contribute to the project? (contributions will go through us not edX)

File: docs/getting_started.rst [hunk 5]
```diff
@@ -7,10 +7,61 @@ below is executed within the virtualenv.
...
 --------------------
-Dependencies can be installed via the command below.
+
+Dependencies can be installed via `pip` in your project's virtual environment:
+
+.. code-block:: bash
+
+    (venv) $ pip install fonzie
+
+Alternatively, if you indend to work on this project, clone the repository first and then
```
Comment (sampaccoud) [inline docs/getting_started.rst]: s/idend/intend

File: docs/getting_started.rst [hunk 6]
```diff
@@ -7,10 +7,61 @@ below is executed within the virtualenv.
...
+install requirements _via_ the command below:
 
 .. code-block:: bash
 
-    $ make requirements
+    $ git clone git@github.com:openfun/fonzie.git
+    (venv) $ make requirements
```
Comment (sampaccoud) [inline docs/getting_started.rst:41]: venv creation being implicit

File: docs/getting_started.rst [hunk 7]
```diff
@@ -7,10 +7,61 @@ below is executed within the virtualenv.
...
+    INSTALLED_APPS = (
+        # [...]
+        'fonzie',
+    )
+
+    # Django Rest Framework (aka DRF)
+    REST_FRAMEWORK = {
+        'ALLOWED_VERSIONS': ('1.0', ),
+        'DEFAULT_VERSION': '1.0',
+        'DEFAULT_VERSIONING_CLASS': 'rest_framework.versioning.URLPathVersioning',
```
Comment (sampaccoud) [inline docs/getting_started.rst:63]: header version from day one would be cool, even if we lose openAPI automatic schema?

File: docs/internationalization.rst [hunk 8]
```diff
@@ -2,25 +2,31 @@
...
+translation community will translate the content, after which we can retrieve
+the translations.
 
 .. _Transifex: https://www.transifex.com/
 
-Pushing source translation files to Transifex requires access to the edx-platform. Request access from the Open Source
-Team if you will be pushing translation files. You should also `configure the Transifex client`_ if you have not done so
-already.
+Pushing source translation files to Transifex requires access to the
+edx-platform. Request access from the Open Source Team if you will be pushing
```
Comment (sampaccoud) [inline docs/internationalization.rst]: not sure it applies to us

File: docs/internationalization.rst [hunk 9]
```diff
@@ -39,13 +45,16 @@ The `make` targets listed below can be used to push or pull translations.
...
 ~~~~~~~~~~~~~~~~~
-As you develop features it may be helpful to know which strings have been marked for translation, and which are not.
-Use the `fake_translations` make target for this purpose. This target will extract all strings marked for translation,
-generate fake translations in the Esperanto (eo) language directory, and compile the translations.
 
-You can trigger the display of the translations by setting your browser's language to Esperanto (eo), and navigating to
-a page on the site. Instead of plain English strings, you should see specially-accented English strings that look
-like this:
+As you develop features it may be helpful to know which strings have been marked
+for translation, and which are not. Use the `fake_translations` make target for
```
Comment (sampaccoud) [inline docs/internationalization.rst]: s/are not/have not

File: fonzie/urls/__init__.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+from django.conf.urls import include, url
+
+from ..apps import FonzieConfig
+from . import schema as schema_urls
+from . import status as status_urls
+
+app_name = FonzieConfig.name
+
+API_PREFIX = r'^v(?P<version>[0-9]+\.[0-9]+)'
```
Comment (sampaccoud) [inline fonzie/urls/__init__.py]: header versioning blablabla

File: requirements/dev.txt [hunk 11]
```diff
@@ -6,29 +6,34 @@
...
-backports.functools-lru-cache==1.5  # via astroid, caniusepython3, pylint
+backports.functools-lru-cache==1.5  # via caniusepython3
 caniusepython3==6.0.0
 certifi==2018.1.18        # via requests
 chardet==3.0.4            # via requests
 click-log==0.1.8          # via edx-lint
-click==6.7                # via click-log, edx-lint, pip-tools
-configparser==3.5.0       # via pydocstyle, pylint
+click==6.7                # via click-log, drf-openapi, edx-lint, pip-tools
+coreapi==2.3.3            # via django-rest-swagger, openapi-codec
```
Comment (sampaccoud) [inline requirements/dev.txt]: will not be necessary if you use the same version of DRF as edX ?

File: requirements/doc.txt [hunk 12]
```diff
@@ -9,28 +9,37 @@ babel==2.5.3              # via sphinx
 bleach==2.1.2             # via readme-renderer
 certifi==2018.1.18        # via requests
 chardet==3.0.4            # via doc8, requests
-django==1.11.10
+click==6.7                # via drf-openapi
+coreapi==2.3.3            # via django-rest-swagger, openapi-codec
+coreschema==0.0.4         # via coreapi
+django-rest-swagger==2.1.2  # via drf-openapi
+django==2.0.2
```
Comment (sampaccoud) [inline requirements/doc.txt]: most changes to this file are no more relevant if this app has to run in OpenedX?

File: tests/views/test_schema.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        self.client = RequestsClient()
+
+    def test_get(self):
+        """
+        A GET request should return the API schema in the following formats:
+
+        * coreapi (default)
+        * openapi
+        """
+        print(self.url)
```
Comment (sampaccoud) [inline tests/views/test_schema.py]: print fail

File: tests/views/test_status.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    """Tests for the Versionview"""
+
+    def setUp(self):
+        """Set view url"""
+
+        self.url = reverse('status:version', kwargs={'version': '1.0'})
+
+    def test_get(self):
+        """HTTP/GET returns API version"""
+        print(self.url)
```
Comment (sampaccoud) [inline tests/views/test_status.py]: print fail.

Not a big fan of using the setUp method. it's cheap enough to compute the url in each test  and makes each test self understandable.

File: README.rst (hunk 3 above)
Comment (jmaupetit) [reply README.rst:15]: Not really. See: http://williamdurand.fr/2015/06/02/video-nobody-understands-rest/

File: docs/getting_started.rst (hunk 7 above)
Comment (jmaupetit) [reply docs/getting_started.rst:63]: It's an easy switch. I think we have time to postpone this feature (waiting for [other versioning schema in DRF OpenAPI](https://github.com/limdauto/drf_openapi#4-constraints)).

File: docs/internationalization.rst (hunk 9 above)
Comment (jmaupetit) [reply docs/internationalization.rst]: We have to spread the word to `edx/cookiecutter-django-app` authors!

File: fonzie/urls/__init__.py (hunk 10 above)
Comment (jmaupetit) [reply fonzie/urls/__init__.py]: I've added a comment for that.

File: tests/views/test_schema.py (hunk 13 above)
Comment (jmaupetit) [reply tests/views/test_schema.py]: :fearful:

File: tests/views/test_status.py (hunk 14 above)
Comment (jmaupetit) [reply tests/views/test_status.py]: It's not really DRY if you test a single endpoint in a `TestCase`.

File: requirements/doc.txt (hunk 12 above)
Comment (jmaupetit) [reply requirements/doc.txt]: FTR, we will stick to Open edX Hawthorn Django and DRF releases constraints.

File: README.rst (hunk 3 above)
Comment (sampaccoud) [reply README.rst:15]: bikeshedding FTW ^^

File: tests/views/test_status.py (hunk 14 above)
Comment (sampaccoud) [reply tests/views/test_status.py]: I don't think tests should be DRY. It's more important that they be self contained, simple and fast.
What you do in setup is done before EACH test in your TestCase is played. So by experience:
- you end up doing things that only a fraction of your tests need (often adding objects to the db which is very slow) and slowing down your tests.
- when the number of tests grows I need to scroll up to see what you're doing to understand each test. Painful.
- hard to maintain when adding/deleting tests.
- hard to maintain when what a new test needs is nearly what you're doing in setup but not quite... so people start recreating it in the test or worse modifying in the test what you created in the setup aaaaaahhhh.

Comment (sampaccoud) [reply tests/views/test_status.py]: So IMO setup should only be used exceptionnaly for things that are sure to be needed for all tests in the TestCase now and ever ^^

Comment (jmaupetit) [reply tests/views/test_status.py]: You mean like the `TestCase` target url for example? :trollface:

Comment (jmaupetit) [reply tests/views/test_status.py]: Trolling apart, I do agree with you and I will make a proper usage of the `TestCase.setup` method.

File: tests/views/test_status.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        response = self.client.get(self.url, format='json')
+        assert response.status_code == status.HTTP_200_OK
+        assert response.json() == {'version': fonzie_version}
+
+    def test_post(self):
+        """HTTP/POST should return a 405"""
+
+        response = self.client.post(self.url, data={}, format='json')
+        assert response.status_code == status.HTTP_405_METHOD_NOT_ALLOWED
```
Comment (sampaccoud) [inline tests/views/test_status.py:34]: why not test the other verbs as well?

Comment (jmaupetit) [reply tests/views/test_status.py:34]: I wrote this test to check that DRF is doing its job properly and he does :wink: Hence, we can remove this test instead.

File: fonzie/views/status.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from fonzie import __version__ as fonzie_version
+
+
+class VersionView(APIView):
+    """
+    API endpoint to get the running API version
```
Comment (sampaccoud) [inline fonzie/views/status.py:15]: so here should you write 
"""API endpoint to get the running API version"""
(ref to our talk this morning :-)

**Review by sampaccoud — APPROVED** [review]

File: fonzie/views/status.py (hunk 16 above)
Comment (jmaupetit) [reply fonzie/views/status.py:15]: I'll fix this in another PR.

**Pull request merged**